src/common/eeprom.c

```diff
@@ -2,228 +2,303 @@
-    VARIANT8_UI16, // EEVAR_VERSION
-    VARIANT8_UI8,  // EEVAR_FILAMENT_TYPE
-    VARIANT8_UI32, // EEVAR_FILAMENT_COLOR
-    VARIANT8_FLT,  //
-    VARIANT8_FLT,  //
-    VARIANT8_FLT,  //
-    VARIANT8_UI8,  // EEVAR_RUN_SELFTEST
-    VARIANT8_UI8,  // EEVAR_RUN_XYZCALIB
-    VARIANT8_UI8,  // EEVAR_RUN_FIRSTLAY
-    VARIANT8_UI8,  // EEVAR_FSENSOR_ENABLED
-    VARIANT8_UI8,  // EEVAR_LAN_FLAG
-    VARIANT8_UI32, // EEVAR_LAN_IP4_ADDR    X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_MSK     X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_GW      X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_DNS1    X.X.X.X address encoded in uint32
-    VARIANT8_UI32, // EEVAR_LAN_IP4_DNS2    X.X.X.X address encoded in uint32
-    VARIANT8_UI8,  // EEVAR_LAN_HOSTNAME_START   Start of 20char string
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
-    VARIANT8_UI8,
+#include "cmsis_os.h"
+#include "ff.h"
+#include "crc32.h"
+
+#define EEPROM_VARCOUNT (sizeof(eeprom_map) / sizeof(eeprom_entry_t))
+#define EEPROM_DATASIZE sizeof(eeprom_vars_t)
+#define EEPROM__PADDING 3
+
+#define EEPROM_MAX_NAME          16     // maximum name length (with '\0')
+#define EEPROM_MAX_DATASIZE      256    // maximum datasize
+#define EEPROM_FIRST_VERSION_CRC 0x0004 // first eeprom version with crc support
+
+// flags will be used also for selective variable reset default values in some cases (shipping etc.))
+#define EEVAR_FLG_READONLY 0x0001 // variable is read only
+
+// measure time needed to update crc
+//#define EEPROM_MEASURE_CRC_TIME
+
+#pragma pack(push)
+#pragma pack(1)
+
+// eeprom map entry structure
+typedef struct _eeprom_entry_t {
+    const char name[EEPROM_MAX_NAME];
+    uint8_t type;   // variant8 data type
+    uint8_t count;  // number of elements
+    uint16_t flags; // flags
+} eeprom_entry_t;
+
+// eeprom vars structure (used for defaults)
+typedef struct _eeprom_vars_t {
+    uint16_t VERSION;
+    uint16_t FEATURES;
+    uint16_t DATASIZE;
+    uint16_t FWVERSION;
+    uint16_t FWBUILD;
+    uint8_t FILAMENT_TYPE;
+    uint32_t FILAMENT_COLOR;
+    uint8_t RUN_SELFTEST;
+    uint8_t RUN_XYZCALIB;
+    uint8_t RUN_FIRSTLAY;
+    uint8_t FSENSOR_ENABLED;
+    float ZOFFSET;
+    float PID_NOZ_P;
+    float PID_NOZ_I;
+    float PID_NOZ_D;
+    float PID_BED_P;
+    float PID_BED_I;
+    float PID_BED_D;
+    uint8_t LAN_FLAG;
+    uint32_t LAN_IP4_ADDR;
+    uint32_t LAN_IP4_MSK;
+    uint32_t LAN_IP4_GW;
+    uint32_t LAN_IP4_DNS1;
+    uint32_t LAN_IP4_DNS2;
+    char LAN_HOSTNAME[LAN_HOSTNAME_MAX_LEN + 1];
+    char _PADDING[EEPROM__PADDING];
+    uint32_t CRC32;
+} eeprom_vars_t;
```

Comment:
EEPROM_FEATURES is reflected only in header, why it should be reflected here and how?
Adding variables to eeprom is not every day job. EEprom structure and table will be checked in eeprom unit test, because it cannot be simply checked in compile time. Only padding vs datasize can be checked with static_assert - this will be done with C++ eeprom encapsulation. Some runtime check can be done inside init, i will add this later.